### DISCUSSION TITLE
importing functions from '@react-native-firebase/functions' always fails

### DISCUSSION THREAD

**Discussion by tycali17** [invertase/react-native-firebase#4654] [Q&A]
Whenever I run any project with that import, it fails with 'native module cannot be null'. React Native clearly auto linked them, I've tried re-installing, manually linking to iOS, and so much more. I cannot for the life of me get my app running without errors and thus cannot use firebase functions in react native. How can I ensure the import stops breaking the app? I even see the node module folder 'functions' with content inside. I used yarn add for RNFBApp & RNFBFunctions

<img width="490" alt="Screen Shot 2020-12-05 at 8 57 36 AM" src="https://user-images.githubusercontent.com/22963509/101247647-f880ef00-36d7-11eb-8aa5-fe56d34746af.png">

<img width="721" alt="Screen Shot 2020-12-05 at 8 58 13 AM" src="https://user-images.githubusercontent.com/22963509/101247669-0b93bf00-36d8-11eb-82fb-2e0a8ff38b73.png">

I hope someone can assist, I've spent literally several months on this issue and read maybe 100 posts so far.

**Comment by mikehardy** [comment]
Have you tried examining a working app? This will create one, then you can see it run (which serves two purposes: proves it *can* work which is sometimes needed to maintain the courage to continue on, and gives you something to compare with your app so hopefully you can find the problem)  https://github.com/mikehardy/rnfbdemo/blob/master/make-demo.sh

**Reply by tycali17** [reply]
I haven't tried this -- I'll give it a shot, thanks! Will update on results

**Reply by tycali17** [reply]
> Have you tried examining a working app? This will create one, then you can see it run (which serves two purposes: proves it _can_ work which is sometimes needed to maintain the courage to continue on, and gives you something to compare with your app so hopefully you can find the problem) https://github.com/mikehardy/rnfbdemo/blob/master/make-demo.sh

Thanks for the help. So I figured out how to stop the error by using the demo and filling in the gaps in my own app.

I'm not sure if I need to make a new thread, but after deploying the helloWorld function and trying to get the response.data from it, I'm met with the error "Error: The data couldn’t be read because it isn’t in the correct format."

Is this usually on the backend function's end, or is it referring to how I'm grabbing the data on the frontend?

Here's backend function 'helloWorld' on firebase:
<img width="649" alt="Screen Shot 2020-12-16 at 11 53 08 AM" src="https://user-images.githubusercontent.com/22963509/102393283-627d7c00-3f95-11eb-89c1-4114df22e50e.png">

And the code on frontend getting data: (added correct imports)
<img width="429" alt="Screen Shot 2020-12-16 at 11 54 57 AM" src="https://user-images.githubusercontent.com/22963509/102393385-88a31c00-3f95-11eb-827f-e44c9278ff04.png">

**Reply by mikehardy** [reply]
I can only refer you to the documentation, which has a pretty thorough walk through plus ways to test the functions themselves from the command line https://rnfirebase.io/functions/writing-deploying-functions

**Reply by tycali17** [reply]
thanks, I've read the docs so many times but will keep practicing. I'm at least on the right track finally. Thanks again

**Reply by mikehardy** [reply]
In case it helps, I have a generic "sendSuccess" function that I call in all my functions to handle some log correlation header stuff but also to wrap the result so that it is in the correct format, it looks like this in typescript and I call it with the "res" object you are passed to start the function, and any arbitrary object that can be JSON stringified:

```typescript

export function sendSuccess(
  res: functions.Response,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  body: { result: any }
): functions.Response {
  console.log(
    Date.now() +
      ' Success Correlation-ID: "' +
      res.locals.correlationId +
      '" kullkiId: "' +
      res.locals.kullkiId +
      '" body: "' +
      JSON.stringify(body) +
      '"'
  );
  const response = res
    .set('X-Correlation-ID', res.locals.correlationId)
    .send(JSON.stringify({ data: body }) + '\n');
  return response;
}
```

I use command-line curl to test these things while developing, and then when I first hook them up in an app I usually console.log(JSON.stringify(theResponseObjectIGot)) in order to make sure my object shape is what I expected, and I very slowly inch forward until I have a new feature done

Hope this helps

**Reply by tycali17** [reply]
It does, very much. Will certainly give this a go along with curl testing. Many thanks

**Reply by tycali17** [reply]
So I've had success up to the point where I have to call the function that I put on firebase. It's now saying I need to link the functions() module or add to podfile after I used yarn add + pod install etc.

Do you suggest basically continuing to comb through the demo and compare files with my project to ensure it works properly? Also noticed @FirebaseSdkVersion is set in demo but it runs me into version conflicts for the related firebase dependencies. Not sure if I need the $FirebaseSDKVersion set or not.

If I should just keep comparing projects I will do so, thanks in advance for any help.

**Reply by mikehardy** [reply]
If the demo works and your project doesn't, either continue combing through the demo or do a large scale harvest-logic-from-project / drop-it-in-working-demo replace of the broken project

**Comment by tycali17** [comment]
Thanks, will do. Sent from Mail for Windows 10 From: Mike HardySent: Monday, December 28, 2020 7:31 AMTo: invertase/react-native-firebaseCc: tycali17; AuthorSubject: Re: [invertase/react-native-firebase] importing functions from '@react-native-firebase/functions' always fails (#4654) If the demo works and your project doesn't, either continue combing through the demo or do a large scale harvest-logic-from-project / drop-it-in-working-demo replace of the broken project—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.
